backend/alert/alerts.py

```diff
@@ -48,46 +49,134 @@ def __init__(self, template, reg):
-                html=self.text,
+                subject=alert_subject,
+                html=alert_text,
             )
         except SMTPRecipientsRefused:
             logger.exception("Email Error")
-            return False
+            return None
 
 
 class Text(Alert):
     def __init__(self, reg):
-        super().__init__("alert/text_alert.txt", reg)
-
-    def send_alert(self):
+        super().__init__("alert/text_alert.txt", reg, close_template="alert/text_alert_close.txt")
+
+    def send_alert(self, close_notification=False):
+        """
+        Returns False if notification was not sent intentionally,
+        and None if notification was attempted to be sent but an error occurred.
+        """
+        if (
+            self.registration.user
+            and self.registration.user.profile
+            and self.registration.user.profile.push_notifications
+        ):
+            # Do not send text if push_notifications is enabled
+            return False
         if self.registration.user is not None and self.registration.user.profile.phone is not None:
             phone_number = self.registration.user.profile.phone
         elif self.registration.phone is not None:
             phone_number = self.registration.phone
         else:
             return False
 
-        return send_text(phone_number, self.text)
+        if close_notification:
+            if not self.close_text:
+                # This should be unreachable
+                return None
+            alert_text = self.close_text
+        else:
+            alert_text = self.text
+        return send_text(phone_number, alert_text)
+
+
+class PushNotification(Alert):
+    def __init__(self, reg):
+        super().__init__("alert/push_notif.txt", reg, close_template="alert/push_notif_close.txt")
+
+    def send_alert(self, close_notification=False):
+        """
+        Returns False if notification was not sent intentionally,
+        and None if notification was attempted to be sent but an error occurred.
+        """
+        if (
+            self.registration.user
+            and self.registration.user.profile
+            and self.registration.user.profile.push_notifications
+        ):
+            # Only send push notification if push_notifications is enabled
+            pennkey = self.registration.user.username
+            bearer_token = str(10101)
+            if close_notification:
+                if not self.close_text:
+                    # This should be unreachable
+                    return None
+                alert_title = "%s just closed." % self.registration.section.full_code
+                alert_body = self.close_text
+            else:
+                alert_title = "%s is now open!" % self.registration.section.full_code
+                alert_body = self.text
+            try:
+                response = requests.post(
+                    "https:/api.pennlabs.org/notifications/send/internal",
```

Comment:
Not sure if this needs to be in this PR, but we should probably hunt around for all of the hardcoded URLs and put them in a `constants.py` file or something

Reply:
Agreed